File: hackerrank/problem solving/algorithms/2 - implementation/33 - library fine/library_fine.js
// https://www.hackerrank.com/challenges/library-fine/problem?isFullScreen=true

const HackosFine = Object.freeze({
  DaysFine: 15,
  MonthsFine: 500,
  YearsFine: 10000,
});

class DateThatWorks {
  #day;
  #month;
  #year;

  constructor(day, month, year) {
    this.#day = day;
    this.#month = month;
    this.#year = year;
  }

  day() {
    return this.#day;
  }

  month() {
    return this.#month;
  }

  year() {
    return this.#year;
  }
}

function main() {
  const returnDate = readDate();
  const dueDate = readDate();
  console.log(calculateFine(returnDate, dueDate));
}

function readDate() {
  const [day, month, year] = readLine().split(' ').map(x => +x);
  return new DateThatWorks(day, month, year);
}

// T: O(1)
// S: O(1) extra space
function calculateFine(returnDate, dueDate) {
  if (isYearLate(returnDate, dueDate)) return HackosFine.YearsFine;
  if (isMonthLate(returnDate, dueDate))
    return (returnDate.month() - dueDate.month()) * HackosFine.MonthsFine;
  if (isDayLate(returnDate, dueDate))
    return (returnDate.day() - dueDate.day()) * HackosFine.DaysFine;
  return 0;
}

function isYearLate(returnDate, dueDate) {
  return returnDate.year() > dueDate.year();
}

function isMonthLate(returnDate, dueDate) {
  return (
    returnDate.year() === dueDate.year() && returnDate.month() > dueDate.month()
  );
}

function isDayLate(returnDate, dueDate) {
  return (
    returnDate.year() === dueDate.year() &&
    returnDate.month() === dueDate.month() &&
    returnDate.day() > dueDate.day()
  );
}

////////////////////////////////////////////////

const readline = require('readline');

const rl = readline.createInterface({ input: process.stdin });
let inputLines = [];
rl.on('line', (line) => inputLines.push(line));
rl.on('close', main);

function readLine() {
  return inputLines.shift();
}

File: hackerrank/problem solving/algorithms/2 - implementation/33 - library fine/library_fine.cs
// https://www.hackerrank.com/challenges/library-fine/problem?isFullScreen=true

DateTime returnDate = ReadDate();
DateTime dueDate = ReadDate();
Console.WriteLine(CalculateFine(returnDate, dueDate));

DateTime ReadDate()
{
    var arr = Console.ReadLine()!.Split().Select(int.Parse).ToArray();
    int day = arr[0];
    int month = arr[1];
    int year = arr[2];
    return new DateTime(year, month, day);
}

// T: O(1)
// S: O(1) extra space
int CalculateFine(DateTime returnDate, DateTime dueDate)
{
    if (IsYearLate(returnDate, dueDate))
        return (int)HackosFine.YearsFine;
    if (IsMonthLate(returnDate, dueDate))
        return (returnDate.Month - dueDate.Month) * (int)HackosFine.MonthsFine;
    if (IsDayLate(returnDate, dueDate))
        return (returnDate.Day - dueDate.Day) * (int)HackosFine.DaysFine;
    return 0;
}

bool IsYearLate(DateTime returnDate, DateTime dueDate)
{
    return returnDate.Year > dueDate.Year;
}

bool IsMonthLate(DateTime returnDate, DateTime dueDate)
{
    return returnDate.Year == dueDate.Year && returnDate.Month > dueDate.Month;
}

bool IsDayLate(DateTime returnDate, DateTime dueDate)
{
    return returnDate.Year == dueDate.Year && returnDate.Month == dueDate.Month && returnDate.Day > dueDate.Day;
}

enum HackosFine
{
    DaysFine = 15,
    MonthsFine = 500,
    YearsFine = 10000
}
